src/marker_group.js

```diff
@@ -0,0 +1,88 @@
+"use strict";
+
+/*
+Potential improvements:
+- cap the length of rendered marker range
+- save rendered markers and only search through rendered markers when looking for hover match
+- don't cut off total amount of markers at 500
+*/
+
+class MarkerGroup {
+    constructor(session) {
+        this.markers = [];
+        this.session = session;
+        session.addDynamicMarker(this);
+    }
+
+    /**
+     * Finds the first marker containing pos
+     * @param {Position} pos 
+     * @returns MarkerGroupItem
+     */
+    getMarkerAtPosition(pos) {
+        return this.markers.find(function(marker) {
+            return marker.range.contains(pos.row, pos.column);
+        });
+    }
+
+    /**
+     * Comparator for Array.sort function, which sorts marker definitions by their positions
+     * 
+     * @param {Ace.TooltipMarker} a first marker.
+     * @param {Ace.TooltipMarker} b second marker.
+     * @returns {number} negative number if a should be before b, positive number if b should be before a, 0 otherwise.
+     */
+    markersComparator(a, b) {
+        return a.range.start.row - b.range.start.row;
+    }
+
+    /**
+     * Sets marker definitions to be rendered. Limits the number of markers at MAX_MARKERS.
+     * @param {Ace.MarkerGroupItem[]} markers an array of marker definitions.
+     */
+    setMarkers(markers) {
+        this.markers = markers.sort(this.markersComparator).slice(0, this.MAX_MARKERS);
+        this.session._signal("changeBackMarker");
```

Comment:
what does changeBackMarker mean? why is this event needed?

Reply:
@nightwing ^^

Reply:
Ace has two sets of markers, one behind and one above the text, this event's let's editor know that markers behind the text have changed and must be rendered again.